ambuda-org/ambuda
# Optimize container image build

### Changes
1. `poetry` with dev and product dependencies classified - WIP
2. create `wheel` package for ambuda 
3. use python slim image as base image and adopt docker multistage build approach
4. remove `make` and other build tools, as much as possible, from runtime 

### Next 
1. Build `prod` a lighter image
2. Prune unused files and build-time caches
3. Arrive at a small essential and sustainable base image

### Test plan
1. Pull docker image
```
$ docker pull ghcr.io/kvchitrapu/ambuda:ambuda-ci
$ $ docker images 
REPOSITORY                  TAG                  IMAGE ID       CREATED          SIZE
ghcr.io/kvchitrapu/ambuda   ambuda-ci            ec5cc9de03a9   25 minutes ago   858MB
```
2. Update docker-compose.yml
```
~/src/ambuda-kvic$ git diff docker-compose.yml 
diff --git a/docker-compose.yml b/docker-compose.yml
index 31f22a9..bcc90e6 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -2,7 +2,7 @@ version: '3.4'
 
 services:
   app:
-    build: .
+    image: ghcr.io/kvchitrapu/ambuda:ambuda-ci
     tty: true
     volumes:
       - .:/app
@@ -20,7 +20,7 @@ services:
       - AMBUDA_BOT_PASSWORD=insecure bot password
 
   celery:
-    build: .
+    image: ghcr.io/kvchitrapu/ambuda:ambuda-ci
     command: /venv/bin/celery -A ambuda.tasks worker --loglevel=INFO
     volumes:
```
3. Run 
    * On Linux, run  `docker-compose up`
    * On Mac, run `docker compose up`
4. Result
```
app_1     | [1] 
app_1     | [1] Rebuilding...
app_1     | [0]  * [OK] Ambuda database check has passed.
app_1     | [0] [2022-10-10 12:31:56,102] INFO in _internal:  * Running on all addresses (0.0.0.0)
app_1     | [0]    WARNING: This is a development server. Do not use it in a production deployment.
app_1     | [0]  * Running on http://127.0.0.1:5000
app_1     | [0]  * Running on http://172.18.0.3:5000 (Press CTRL+C to quit)
app_1     | [0] [2022-10-10 12:31:56,117] INFO in _internal:  * Restarting with watchdog (inotify)
```

### Note
Not sure why GH picks up my commits all the way from August. Looking to move
